contributors/design-proposals/pod-resource-management.md

```diff
@@ -1,114 +1,367 @@
+is given the lowest priority.
+
+The `Burstable` cgroup sandbox cpu share allocation is dynamic based
+on the set of pods currently scheduled to the node.  
+
+```
+ROOT/Burstable/cpu.shares = max(sum cpu requests of all Burstable pods, 2)
+```
 
-## Design
+The `kubelet` will have a control loop that dynamically adjusts the
+value of the CPU share based on its view of Burstable pods in a
+non-terminal state.  This approach is chosen to reduce overall code
+complexity in the `kubelet`.
 
-High level requirements for the design are as follows:
- - Do not break existing users. Ideally, there should be no changes to the Kubernetes API semantics.
- - Support multiple cgroup managers - systemd, cgroupfs, etc.
+### Memory overcommitment
 
-How we intend to achieve these high level goals is covered in greater detail in the Implementation Plan.
+The `kubelet` highly prioritizes resource utilization, and thus allows
+BestEffort and Burstable pods to potentially consume as many resources
+that are presently available on the node.  In cases where resources become
+scarce, the `kubelet` configures the `oom_killer` and/or pod evictions to
+evict workloads that have the lowest QoS class.
 
-We use the following denotations in the sections below:
+There is latency involved in the pod eviction process which can be a
+cause of concern in latency sensitive servers.  On such servers, we
+want to avoid OOM conditions, and promote the system to steer memory
+consumption on the node to align with the actual amount of
+requested resources for non-terminal Guaranteed and Burstable pods.
 
-For the three QoS classes
-`G⇒ Guaranteed QoS, Bu⇒ Burstable QoS, BE⇒ BestEffort QoS`
+Operators can configure a desired qos memory overcommit (qmo) range using 
+the `--qos-memory-overcommit` flag.  This flag is not enabled
+by default, and operators must opt-in to this behavior.
+
+If enabled, the `kubelet` will dynamically update the Burstable and
+BestEffort cgroup sandbox memory limits in an effort to satisfy the
+specified level of overcommit.
+
+* The flag will accept values in a range from 0-100.
+* A value of 100 will instruct the `kubelet` to adjust the Burstable and
+BestEffort cgroup sandboxes to allow maximal overcommit.
+* A value of 0 will instruct the `kubelet` to adjust the Burstable and
+BestEffort cgroup to restrict memory overcommit by inducing memory
+pressure and reclaim if needed at the QoS level cgroup tiers before
+inducing pressure at the `ROOT` cgroup.
+
+The cgroup sandboxes are adjusted as follows:
+
+```
+ROOT/Burstable/memory.limit_in_bytes = 
+    Node.Allocatable - {(summation of memory requests of `Guaranteed` pods)*(1-qmo/100)}
+ROOT/BestEffort/memory.limit_in_bytes = 
+    Node.Allocatable - {(summation of memory requests of all `Guaranteed` and `Burstable` pods)*(1-qmo/100)}
+```
+
+Each time a pod is admitted to the `kubelet`, the `kubelet` will increment
+the QoS level cgroup memory limits.  In order to safely reduce memory limits, 
```

Comment:
Why are we doing QoS cgroups just for memory? If we are implementing it, we should add support for CPU too.